R/evaluate_projection.r

```diff
@@ -0,0 +1,215 @@
+#'
+#' If an end date is provided, projections will be provided through the
+#' closest expected reporting date to the end date, even if this extrapolates
+#' beyond the end of the provided data set.
+#'
+#' UPDATES FOR THE FUTURE:
+#' - Model must incorporate year-to-year variability
+#' - Model must incorporate state-to-state variability
+#' - Multiple models can be provided, with scaling, to pool projections
+#'
+#' @description Evaluate a model that projects uptake into the future
+#'
+#' @param old_data uptake data used to fit the projection model
+#' @param model model fit to project uptake into the future
+#' @param new_data uptake data to compare the projection model against
+#' @param start_date which date to perform projection from, in %Y-%m-%d format
+#' @param end_date which date to perform projections until, in %Y-%m-%d format
+#' @param conf width of credible interval to plot
+#'
+#' @return data frame of the data underlying diagnostic plots
+#'
+#' @export
+#'
+evaluate_projection <- function(old_data, model, new_data = NULL,
+                                start_date = NULL, end_date = NULL,
+                                conf = 0.95) {
+
+  model_rows <- (nrow(old_data) - nrow(model$data) + 1):(nrow(old_data))
+
+  daily_mean <- mean(old_data$daily[model_rows])
+  daily_sd <- sd(old_data$daily[model_rows])
+  previous_mean <- mean(old_data$daily[model_rows - 1])
+  previous_sd <- sd(old_data$daily[model_rows - 1])
+  elapsed_mean <- mean(old_data$elapsed[model_rows])
+  elapsed_sd <- sd(old_data$elapsed[model_rows])
+
+  if (is.null(new_data)) {
+    output <- old_data
+  } else {
+    output <- new_data
+  }
+
+  output$previous <- c(NA, output$daily[1:(nrow(output) - 1)])
+  output$interval <- c(output$elapsed[1], diff(output$elapsed))
+
+  if (!is.null(end_date)) {
+    end_date <- as.Date(end_date)
+    if (end_date <= output$date[nrow(output)]) {
+      output <- output[output$date <= end_date, ]
+    } else {
+      extra_dates <- seq(output$date[nrow(output)], end_date,
+                         by = round(mean(output$interval)))[-1]
+      extra_rows <- nrow(output) + seq_along(length(extra_dates))
+      output[extra_rows, ] <- NA
+      output$date[extra_rows] <- extra_dates
+      output$elapsed <- as.numeric(output$date -
+                                     (output$date[1] - output$elapsed[1]))
+      output$previous[extra_rows[1]] <- output$daily[extra_rows[1] - 1]
+    }
+  }
+
+  output$incident_mean <- rep(NA, nrow(output))
+  output$incident_upper <- rep(NA, nrow(output))
+  output$incident_lower <- rep(NA, nrow(output))
+  output$cumulative_mean <- rep(NA, nrow(output))
+  output$cumulative_upper <- rep(NA, nrow(output))
+  output$cumulative_lower <- rep(NA, nrow(output))
+
+  if (!is.null(start_date)) {
+    start_date <- as.Date(start_date)
+    early_dates <- which(output$date <= start_date | is.na(output$previous))
+  } else {
+    early_dates <- which(is.na(output$previous))
```

Comment:
So `early_dates` is not actual date but index. And the entire `output` will be used for evaluation, i.e, the `start_date` will be the start of the `output`, if there is no `start_date`. It may help to read if adding some comments there.

Reply:
Yes, I apologize that this is sloppy in the same way that many of @swo's comments point out sloppy or inefficient things!